Unable to retrieve Tasks when one specific Task has been executed due to a missing class (com.vmware.vim25.ArrayOfDVSOpaqueCommandResultInfo)

See below. When a need for this class is encountered, I can't retrieve any Tasks. Any ways to capture that exception differently to continue the processing of the other Tasks?

18/09/05 16:09:32 ERROR root: ClassNotFoundException caught for type: ArrayOfDVSOpaqueCommandResultInfo
java.lang.ClassNotFoundException: com.vmware.vim25.ArrayOfDVSOpaqueCommandResultInfo
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at org.doublecloud.ws.util.TypeUtil.getVimClass(TypeUtil.java:100)
	at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:280)
	at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:356)
	at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:363)
	at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:356)
	at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:233)
	at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:124)
	at com.vmware.vim25.ws.SoapClient.unMarshall(SoapClient.java:256)
	at com.vmware.vim25.ws.WSClient.invoke(WSClient.java:96)
	at com.vmware.vim25.ws.VimStub.retrieveProperties(VimStub.java:106)
	at com.vmware.vim25.mo.PropertyCollector.retrieveProperties(PropertyCollector.java:98)
	at com.vmware.vim25.mo.ManagedObject.retrieveObjectProperties(ManagedObject.java:146)
	at com.vmware.vim25.mo.ManagedObject.getCurrentProperty(ManagedObject.java:167)
	at com.vmware.vim25.mo.TaskHistoryCollector.getLatestPage(TaskHistoryCollector.java:56)
	at de.net.ScloudAudit.ScloudAuditHostTasksDetails.audit(ScloudAuditHostTasksDetails.java:109)
	at de.net.ScloudAudit.ScloudAudit.audit(ScloudAudit.java:44)
	at de.net.ScloudAudit.ScloudAudit.main(ScloudAudit.java:100)
java.lang.RuntimeException: java.rmi.RemoteException: Exception in SoapClient.invoke:; nested exception is: 
	java.lang.NullPointerException
	at com.vmware.vim25.mo.ManagedObject.retrieveObjectProperties(ManagedObject.java:149)
	at com.vmware.vim25.mo.ManagedObject.getCurrentProperty(ManagedObject.java:167)
	at com.vmware.vim25.mo.TaskHistoryCollector.getLatestPage(TaskHistoryCollector.java:56)
	at de.net.ScloudAudit.ScloudAuditHostTasksDetails.audit(ScloudAuditHostTasksDetails.java:109)
	at de.net.ScloudAudit.ScloudAudit.audit(ScloudAudit.java:44)
	at de.net.ScloudAudit.ScloudAudit.main(ScloudAudit.java:100)
Caused by: java.rmi.RemoteException: Exception in SoapClient.invoke:; nested exception is: 
	java.lang.NullPointerException
	at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:127)
	at com.vmware.vim25.ws.SoapClient.unMarshall(SoapClient.java:256)
	at com.vmware.vim25.ws.WSClient.invoke(WSClient.java:96)
	at com.vmware.vim25.ws.VimStub.retrieveProperties(VimStub.java:106)
	at com.vmware.vim25.mo.PropertyCollector.retrieveProperties(PropertyCollector.java:98)
	at com.vmware.vim25.mo.ManagedObject.retrieveObjectProperties(ManagedObject.java:146)
	... 5 more

Hi Michael,

Sorry for the delay. Was out of commission for a while!
The “test” code is at the end… Nothing special about it (as it mostly came from existing examples)! Version is 6.0.05
<dependency>
<groupId>com.toastcoders</groupId>
<artifactId>yavijava</artifactId>
<version>6.0.05</version>
</dependency>

Just provide a valid ServiceInstance to the get_Data method…
In our environment, we get the exception below while executing the following line: tis = thc.readNextTasks(50);

Note: we use vCloud Director and NSX and the error happens while using multiple versions of both of them (and on top of both vSphere 6.0 and 6.5). It’s obviously related to DVS tasks.

Regards,
Marc


19/06/24 10:03:35 ERROR util.TypeUtil: ClassNotFoundException caught for type: ArrayOfDVSOpaqueCommandResultInfo
java.lang.ClassNotFoundException: com.vmware.vim25.ArrayOfDVSOpaqueCommandResultInfo
at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
at java.lang.Class.forName0(Native Method)
at java.lang.Class.forName(Class.java:264)
at org.doublecloud.ws.util.TypeUtil.getVimClass(TypeUtil.java:100)
at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:280)
at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:233)
at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:124)
at com.vmware.vim25.ws.SoapClient.unMarshall(SoapClient.java:253)
at com.vmware.vim25.ws.WSClient.invoke(WSClient.java:96)
at com.vmware.vim25.ws.VimStub.readNextTasks(VimStub.java:1930)
at com.vmware.vim25.mo.TaskHistoryCollector.readNextTasks(TaskHistoryCollector.java:60)
at local.net.sptest.app.cloud_app.VSphereB15.get_Data(VSphereB15.java:98)
at local.net.sptest.app.cloud_app.vSpherevCloudVM.main(vSpherevCloudVM.java:138)
Exception in thread "main" java.rmi.RemoteException: Exception in SoapClient.invoke:; nested exception is:
java.lang.NullPointerException
at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:127)
at com.vmware.vim25.ws.SoapClient.unMarshall(SoapClient.java:253)
at com.vmware.vim25.ws.WSClient.invoke(WSClient.java:96)
at com.vmware.vim25.ws.VimStub.readNextTasks(VimStub.java:1930)
at com.vmware.vim25.mo.TaskHistoryCollector.readNextTasks(TaskHistoryCollector.java:60)
at local.net.sptest.app.cloud_app.VSphereB15.get_Data(VSphereB15.java:98)
at local.net.sptest.app.cloud_app.vSpherevCloudVM.main(vSpherevCloudVM.java:138)
Caused by: java.lang.NullPointerException
at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:298)
at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:233)
at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:124)
... 6 more


package local.net.sptest.app.cloud_app;

import java.time.temporal.ChronoUnit;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

import com.vmware.vim25.ManagedObjectReference;
import com.vmware.vim25.TaskFilterSpec;
import com.vmware.vim25.TaskFilterSpecByTime;
import com.vmware.vim25.TaskFilterSpecTimeOption;
import com.vmware.vim25.TaskInfo;
import com.vmware.vim25.TaskInfoState;
import com.vmware.vim25.TaskReason;
import com.vmware.vim25.TaskReasonAlarm;
import com.vmware.vim25.TaskReasonSchedule;
import com.vmware.vim25.TaskReasonSystem;
import com.vmware.vim25.TaskReasonUser;
import com.vmware.vim25.mo.Folder;
import com.vmware.vim25.mo.ManagedEntity;
import com.vmware.vim25.mo.ServiceInstance;
import com.vmware.vim25.mo.TaskHistoryCollector;
import com.vmware.vim25.mo.TaskManager;

public class VSphereB15 {

Folder rootFolder = null;

public VSphereB15(Folder rootFolder) {
this.rootFolder = rootFolder;
}

public GregorianCalendar getLatestDate() {
GregorianCalendar datebeg = GregorianCalendar.from(java.time.ZonedDateTime.now().minus(2, ChronoUnit.HOURS));
return datebeg;
}

static String taskReason(TaskReason tr) {
if (tr instanceof TaskReasonAlarm) {
return " <-- Alarm";
} else if (tr instanceof TaskReasonSchedule) {
return " <-- ScheduledTask";
} else if (tr instanceof TaskReasonSystem) {
return " <-- System";
} else if (tr instanceof TaskReasonUser) {
return " <-- User : " + ((TaskReasonUser) tr).getUserName();
}
return "Unknown";
}

@SuppressWarnings("unused")
static void printTaskInfo(TaskInfo ti) {
Calendar calStart = ti.getStartTime();
Date dateStart = calStart == null ? null : calStart.getTime();
Calendar calStop = ti.getCompleteTime();
Date dateStop = calStop == null ? null : calStop.getTime();
System.out.println("Name: " + ti.getName() + ";Key: " + ti.getKey() + ";Entity: " + ti.getEntityName()
+ ";StartTime: " + dateStart + ";EndTime: " + dateStop);
TaskReason tr = ti.getReason();
ManagedObjectReference mor = ti.getEntity();
if (mor != null) {
System.out.println("Reason:" + taskReason(tr) + " Entity Type: " + mor.getType());
}
}

static TaskFilterSpec createTaskFilterSpec(ManagedEntity ent) {
TaskFilterSpec tfs = new TaskFilterSpec();
tfs.setState(new TaskInfoState[] { TaskInfoState.success });
TaskFilterSpecByTime tFilter = new TaskFilterSpecByTime();
Calendar cal = Calendar.getInstance();
cal.roll(Calendar.HOUR, -8);
tFilter.setTimeType(TaskFilterSpecTimeOption.startedTime);
tfs.setTime(tFilter);
return tfs;
}

static void printTaskInfos(TaskInfo[] tis) {
for (int i = 0; tis != null && i < tis.length; i++) {
printTaskInfo(tis[i]);
}
}

public void get_Data(ServiceInstance si) throws Exception {

System.out.println("\n============ Tasks ============");

TaskManager taskMgr = si.getTaskManager();

if (taskMgr != null) {
Folder root = si.getRootFolder();
TaskFilterSpec tfs = createTaskFilterSpec(root);
TaskHistoryCollector thc = taskMgr.createCollectorForTasks(tfs);
thc.setCollectorPageSize(15);
TaskInfo[] tis = thc.getLatestPage();
printTaskInfos(tis);
int total = 0;
while (true) {
tis = thc.readNextTasks(50);
if (tis == null) {
break;
}
total += tis.length;
printTaskInfos(tis);
}
System.out.println("\nTotal number " + "of tasks retrieved:" + total);
thc.destroyCollector();
}

System.out.println("\nEND");
}
}



On May 16, 2019, at 6:19 PM, Michael Rice <notifications@github.com<mailto:notifications@github.com>> wrote:


How can I recreate this error? What version of vsphere are you using, can you provide some sample code I could use?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/yavijava/yavijava/issues/265?email_source=notifications&email_token=AKIDOJCZV24NNXHHAQO7B23PVWCSPA5CNFSM4FTMFBQ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVSKQCI#issuecomment-493135881>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKIDOJHNYEYCU5JMUMS7N7DPVWCSPANCNFSM4FTMFBQQ>.

Hi Michael,
I played a little more with this, and found out that this task is related to a host entering maintenance mode in all our environments (all configured with NSX and vCloud).
See attached screenshot.
The Task “Execute opaque command for set of entities” seems to be the issue!
Thanks,
Marc.



[cid:8611CBFB-E118-4A51-BECE-5BBA726C304E@net.local]


On Jun 24, 2019, at 10:21 AM, Marc Planquart <Marc.Planquart@tech.net.de<mailto:Marc.Planquart@tech.net.de>> wrote:

Hi Michael,

Sorry for the delay. Was out of commission for a while!
The “test” code is at the end… Nothing special about it (as it mostly came from existing examples)! Version is 6.0.05
<dependency>
<groupId>com.toastcoders</groupId>
<artifactId>yavijava</artifactId>
<version>6.0.05</version>
</dependency>

Just provide a valid ServiceInstance to the get_Data method…
In our environment, we get the exception below while executing the following line: tis = thc.readNextTasks(50);

Note: we use vCloud Director and NSX and the error happens while using multiple versions of both of them (and on top of both vSphere 6.0 and 6.5). It’s obviously related to DVS tasks.

Regards,
Marc


19/06/24 10:03:35 ERROR util.TypeUtil: ClassNotFoundException caught for type: ArrayOfDVSOpaqueCommandResultInfo
java.lang.ClassNotFoundException: com.vmware.vim25.ArrayOfDVSOpaqueCommandResultInfo
at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
at java.lang.Class.forName0(Native Method)
at java.lang.Class.forName(Class.java:264)
at org.doublecloud.ws.util.TypeUtil.getVimClass(TypeUtil.java:100)
at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:280)
at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:233)
at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:124)
at com.vmware.vim25.ws.SoapClient.unMarshall(SoapClient.java:253)
at com.vmware.vim25.ws.WSClient.invoke(WSClient.java:96)
at com.vmware.vim25.ws.VimStub.readNextTasks(VimStub.java:1930)
at com.vmware.vim25.mo.TaskHistoryCollector.readNextTasks(TaskHistoryCollector.java:60)
at local.net.sptest.app.cloud_app.VSphereB15.get_Data(VSphereB15.java:98)
at local.net.sptest.app.cloud_app.vSpherevCloudVM.main(vSpherevCloudVM.java:138)
Exception in thread "main" java.rmi.RemoteException: Exception in SoapClient.invoke:; nested exception is:
java.lang.NullPointerException
at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:127)
at com.vmware.vim25.ws.SoapClient.unMarshall(SoapClient.java:253)
at com.vmware.vim25.ws.WSClient.invoke(WSClient.java:96)
at com.vmware.vim25.ws.VimStub.readNextTasks(VimStub.java:1930)
at com.vmware.vim25.mo.TaskHistoryCollector.readNextTasks(TaskHistoryCollector.java:60)
at local.net.sptest.app.cloud_app.VSphereB15.get_Data(VSphereB15.java:98)
at local.net.sptest.app.cloud_app.vSpherevCloudVM.main(vSpherevCloudVM.java:138)
Caused by: java.lang.NullPointerException
at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:298)
at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:233)
at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:124)
... 6 more


package local.net.sptest.app.cloud_app;

import java.time.temporal.ChronoUnit;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

import com.vmware.vim25.ManagedObjectReference;
import com.vmware.vim25.TaskFilterSpec;
import com.vmware.vim25.TaskFilterSpecByTime;
import com.vmware.vim25.TaskFilterSpecTimeOption;
import com.vmware.vim25.TaskInfo;
import com.vmware.vim25.TaskInfoState;
import com.vmware.vim25.TaskReason;
import com.vmware.vim25.TaskReasonAlarm;
import com.vmware.vim25.TaskReasonSchedule;
import com.vmware.vim25.TaskReasonSystem;
import com.vmware.vim25.TaskReasonUser;
import com.vmware.vim25.mo.Folder;
import com.vmware.vim25.mo.ManagedEntity;
import com.vmware.vim25.mo.ServiceInstance;
import com.vmware.vim25.mo.TaskHistoryCollector;
import com.vmware.vim25.mo.TaskManager;

public class VSphereB15 {

Folder rootFolder = null;

public VSphereB15(Folder rootFolder) {
this.rootFolder = rootFolder;
}

public GregorianCalendar getLatestDate() {
GregorianCalendar datebeg = GregorianCalendar.from(java.time.ZonedDateTime.now().minus(2, ChronoUnit.HOURS));
return datebeg;
}

static String taskReason(TaskReason tr) {
if (tr instanceof TaskReasonAlarm) {
return " <-- Alarm";
} else if (tr instanceof TaskReasonSchedule) {
return " <-- ScheduledTask";
} else if (tr instanceof TaskReasonSystem) {
return " <-- System";
} else if (tr instanceof TaskReasonUser) {
return " <-- User : " + ((TaskReasonUser) tr).getUserName();
}
return "Unknown";
}

@SuppressWarnings("unused")
static void printTaskInfo(TaskInfo ti) {
Calendar calStart = ti.getStartTime();
Date dateStart = calStart == null ? null : calStart.getTime();
Calendar calStop = ti.getCompleteTime();
Date dateStop = calStop == null ? null : calStop.getTime();
System.out.println("Name: " + ti.getName() + ";Key: " + ti.getKey() + ";Entity: " + ti.getEntityName()
+ ";StartTime: " + dateStart + ";EndTime: " + dateStop);
TaskReason tr = ti.getReason();
ManagedObjectReference mor = ti.getEntity();
if (mor != null) {
System.out.println("Reason:" + taskReason(tr) + " Entity Type: " + mor.getType());
}
}

static TaskFilterSpec createTaskFilterSpec(ManagedEntity ent) {
TaskFilterSpec tfs = new TaskFilterSpec();
tfs.setState(new TaskInfoState[] { TaskInfoState.success });
TaskFilterSpecByTime tFilter = new TaskFilterSpecByTime();
Calendar cal = Calendar.getInstance();
cal.roll(Calendar.HOUR, -8);
tFilter.setTimeType(TaskFilterSpecTimeOption.startedTime);
tfs.setTime(tFilter);
return tfs;
}

static void printTaskInfos(TaskInfo[] tis) {
for (int i = 0; tis != null && i < tis.length; i++) {
printTaskInfo(tis[i]);
}
}

public void get_Data(ServiceInstance si) throws Exception {

System.out.println("\n============ Tasks ============");

TaskManager taskMgr = si.getTaskManager();

if (taskMgr != null) {
Folder root = si.getRootFolder();
TaskFilterSpec tfs = createTaskFilterSpec(root);
TaskHistoryCollector thc = taskMgr.createCollectorForTasks(tfs);
thc.setCollectorPageSize(15);
TaskInfo[] tis = thc.getLatestPage();
printTaskInfos(tis);
int total = 0;
while (true) {
tis = thc.readNextTasks(50);
if (tis == null) {
break;
}
total += tis.length;
printTaskInfos(tis);
}
System.out.println("\nTotal number " + "of tasks retrieved:" + total);
thc.destroyCollector();
}

System.out.println("\nEND");
}
}



On May 16, 2019, at 6:19 PM, Michael Rice <notifications@github.com<mailto:notifications@github.com>> wrote:


How can I recreate this error? What version of vsphere are you using, can you provide some sample code I could use?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/yavijava/yavijava/issues/265?email_source=notifications&email_token=AKIDOJCZV24NNXHHAQO7B23PVWCSPA5CNFSM4FTMFBQ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVSKQCI#issuecomment-493135881>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKIDOJHNYEYCU5JMUMS7N7DPVWCSPANCNFSM4FTMFBQQ>.

--- a/src/main/java/com/vmware/vim25/ArrayOfDVSOpaqueCommandResultInfo.java
+++ b/src/main/java/com/vmware/vim25/ArrayOfDVSOpaqueCommandResultInfo.java
@@ -1,0 +1,17 @@
+package com.vmware.vim25;
+
+public class ArrayOfDVSOpaqueCommandResultInfo {
+    public DVSOpaqueCommandResultInfo[] DVSOpaqueCommandResultInfo;
+
+    public DVSOpaqueCommandResultInfo[] getDVSOpaqueCommandResultInfo() {
+        return this.DVSOpaqueCommandResultInfo;
+    }
+
+    public DVSOpaqueCommandResultInfo getDVSOpaqueCommandResultInfo(int i) {
+        return this.DVSOpaqueCommandResultInfo[i];
+    }
+
+    public void setDVSOpaqueCommandResultInfo(DVSOpaqueCommandResultInfo[] DVSOpaqueCommandResultInfo) {
+        this.DVSOpaqueCommandResultInfo = DVSOpaqueCommandResultInfo;
+    }
+}

--- a/src/main/java/com/vmware/vim25/ArrayOfHostDatastoreSystemDatastoreResult.java
+++ b/src/main/java/com/vmware/vim25/ArrayOfHostDatastoreSystemDatastoreResult.java
@@ -1,0 +1,17 @@
+package com.vmware.vim25;
+
+public class ArrayOfHostDatastoreSystemDatastoreResult {
+    public HostDatastoreSystemDatastoreResult[] HostDatastoreSystemDatastoreResult;
+
+    public HostDatastoreSystemDatastoreResult[] getHostDatastoreSystemDatastoreResult() {
+        return this.HostDatastoreSystemDatastoreResult;
+    }
+
+    public HostDatastoreSystemDatastoreResult getHostDatastoreSystemDatastoreResult(int i) {
+        return this.HostDatastoreSystemDatastoreResult[i];
+    }
+
+    public void setHostDatastoreSystemDatastoreResult(HostDatastoreSystemDatastoreResult[] HostDatastoreSystemDatastoreResult) {
+        this.HostDatastoreSystemDatastoreResult = HostDatastoreSystemDatastoreResult;
+    }
+}

--- a/src/main/java/com/vmware/vim25/DVSOpaqueCommandResultInfo.java
+++ b/src/main/java/com/vmware/vim25/DVSOpaqueCommandResultInfo.java
@@ -1,0 +1,12 @@
+package com.vmware.vim25;
+
+import lombok.Getter;
+import lombok.Setter;
+
+/**
+ * @since SDK5.5
+ */
+public class DVSOpaqueCommandResultInfo extends DynamicData {
+    @Getter @Setter public ManagedObjectReference obj;
+    @Getter @Setter public MethodFault fault;
+}

--- a/src/main/java/com/vmware/vim25/mo/VirtualMachine.java
+++ b/src/main/java/com/vmware/vim25/mo/VirtualMachine.java
@@ -522,6 +522,13 @@ public class VirtualMachine extends ManagedEntity {
     }
 
     /**
+     * @since SDK5.5
+     */
+    public void terminate() throws InvalidState, TaskInProgress, RuntimeFault, RemoteException {
+        getVimService().terminateVM(getMOR());
+    }
+
+    /**
      * @since SDK4.0
      */
     public Task terminateFaultTolerantVM_Task(VirtualMachine vm) throws TaskInProgress, VmFaultToleranceIssue, InvalidState, RuntimeFault, RemoteException {

